articles/zero_copy.md

```diff
@@ -0,0 +1,249 @@
+* Memory operations and copies should be avoided anywhere possible.
+
+## Design Proposal
+
+### RMW API
+
+Introduce APIs for creating/destroying loaned messages, as well as structure for management:
+
+```
+void *
+rmw_allocate_loaned_message(
+  const rmw_publisher_t * publisher,
+  const rosidl_message_type_support_t * type_support,
+  size_t message_size
+);
+
+rmw_ret_t
+rmw_deallocate_loaned_message(
+  const rmw_publisher_t * publisher,
+  void * loaned_message
+);
+```
+
+Extend publisher API for loaned messages:
+
+```
+rmw_ret_t
+rmw_publish(
+  const rmw_publisher_t * publisher
+  const void * ros_message
+  rmw_publisher_allocation_t * allocation,
+  bool is_loaned
+);
+```
+
+Extend subscription API for taking loaned message:
+
+```
+// TODO
+```
+
+### RCL API
+
+Introduce APIs for creating/destroying loaned messages, as well as structure for management:
+
+```
+void *
+rcl_allocate_loaned_message(
+  const rcl_publisher_t * publisher,
+  const rosidl_message_type_support_t * type_support,
+  size_t message_size
+);
+
+rcl_ret_t
+rcl_deallocate_loaned_message(
+  const rcl_publisher_t * publisher,
+  rcl_loaned_message_t * loaned_message
+);
+```
+
+Extend publisher API for loaned messages:
+
+```
+rcl_ret_t
+rcl_publish(
+  const rcl_publisher_t * publisher,
+  const void * ros_message,
+  rmw_publisher_allocation_t * allocation,
+  bool is_loaned
+);
+
+bool
+rcl_publisher_can_loan_messages(const rcl_publisher_t * publisher);
+```
+
+Extend subscription API for taking loaned message:
+
+```
+// TODO
+```
+
+### RCLCPP LoanedMessage
+
+In order to support loaned messages in `rclcpp`, we tntroduce the concept of a `LoanedMessage`.
+A `LoanedMessage` provides a wrapper around the underlying loan mechanisms, and manages the loan's lifecycle.
+
+```
+template <class MsgT, typename Alloc = std::allocator<void>>
+class LoanedMessage
+{
+public:
+  // Get the underlying message
+  MsgT& get()
+
+  // Check if underlying message is valid and consistent
+  bool is_consistent();
+
+  // Loan message from the middleware, if loaning is not available,
+  // allocate memory using Alloc
+  LoanedMessage(
+    const rclcpp::Publisher<MsgT, Alloc> * pub);
+
+private:
+  const rclcpp::Publisher<MsgT, Alloc> * pub_;
+
+  // Holds details of the message loan
+  rcl_loaned_message_t * loaned_msg_;
+
+  // Will be initialized with memory from loaned_msg_, otherwise allocated.
+  std::unique_ptr<MsgT> msg_;
+};
+```
+
+### RCLCPP Publisher
+
+Extend RCLCPP API:
+
+```
+// Return a loaned message from the middleware
+rclcpp::LoanedMessage<MsgT, Alloc>
+rclcpp::Publisher::loan_message()
+
+// Publish a loaned message, returning the loan
+void
+rclcpp::Publisher::publish(std::unique_ptr<rclcpp::LoanedMessage<MsgT, Alloc>> loaned_msg);
+
+// Test if the middleware supports loaning messages
+bool
+rclcpp::Publisher::can_loan_messages()
+```
+
+### RCLCPP Subscription
```

Comment:
I guess, that no modification in RCLCPP subscription would be needed.
Maybe we would need to add a new subscription signature more convenient for loaned messages, but not more than that. Probably, similar changes would be needed in rcl/rmw.